{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import allel\n",
    "import pandas as pd\n",
    "import os\n",
    "from tqdm import tqdm_notebook as tqdm\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "global counter\n",
    "counter=0\n",
    "\n",
    "def make_bed_file(*args):\n",
    "    \n",
    "    patient=args[0]\n",
    "    a=np.empty([1,3])\n",
    "    global bed_pat\n",
    "    global counter\n",
    "    if counter == 0: bed_pat=patient\n",
    "    \n",
    "    try:\n",
    "        try:\n",
    "            sample=args[1]\n",
    "            path_to_files = f\"/media/emir/Storage/Cancer/mutect/rvboost/{patient}s{sample}\"\n",
    "            file_name = f\"intersect.{patient}_s{sample}_0.0.vcf\"\n",
    "            output_name=f\"genes_pat{patient}s{sample}_ensembl.bed\"\n",
    "            callset=allel.read_vcf(os.path.join(path_to_files, file_name),\n",
    "                                   fields=[\"CHROM\",\"POS\"])\n",
    "            for chrom,pos in tqdm(zip(callset[\"variants/CHROM\"], callset[\"variants/POS\"])):\n",
    "                gene_obj=data.genes_at_locus(contig=chrom, position=pos.item())\n",
    "                for obj in gene_obj:\n",
    "                    bed_row=[\"chr\"+str(obj.contig),obj.start,obj.end]\n",
    "                    a=np.vstack([a,bed_row])\n",
    "\n",
    "        except FileNotFoundError:\n",
    "            if bed_pat == patient and counter != 0:\n",
    "                return\n",
    "            else:\n",
    "                path_to_files = f\"/media/emir/Storage/Cancer/mutect/rvboost/{patient}_s1s2\"\n",
    "                file_name = f\"{patient}.rvboost.s1s2.filtered.tsv\"\n",
    "                output_name=f\"genes_pat{patient}_s1s2_ensembl.bed\"\n",
    "                callset = np.loadtxt(fname=os.path.join(path_to_files,file_name),\n",
    "                                     delimiter=\"\\t\", skiprows=1,usecols=(1,2),dtype=str)\n",
    "                for chrom_pos in tqdm(callset):\n",
    "                    contig=str(chrom_pos[0]).split(\"chr\")[1]\n",
    "                    gene_obj=data.genes_at_locus(contig=contig, \n",
    "                                                 position=int(chrom_pos[1]))\n",
    "                    for obj in gene_obj:\n",
    "                        bed_row=[\"chr\"+str(obj.contig),obj.start,obj.end]\n",
    "                        a=np.vstack([a,bed_row])\n",
    "    except OSError:\n",
    "        return\n",
    "        \n",
    "        \n",
    "    a=np.delete(a,(0),axis=0)\n",
    "    df=pd.DataFrame(data=a[0:,0:], columns=[\"CHROM\",\"START\",\"END\"])\n",
    "    df.drop_duplicates(inplace=True)\n",
    "    \n",
    "    arr=df.to_numpy()\n",
    "    arr[:,1:]=arr[:,1:].astype(int)\n",
    "    arr\n",
    "\n",
    "    d={}\n",
    "    for i in arr:\n",
    "        if i[0] not in d.keys():\n",
    "            d[i[0]]=[[i[1],i[2]]]\n",
    "        else:\n",
    "            d[i[0]].append([i[1],i[2]])\n",
    "\n",
    "    with open(os.path.join(path_to_files, output_name),\n",
    "              \"w\") as genes_str:        \n",
    "        for k,v in d.items():\n",
    "            v.sort(key=lambda interval: interval[0])\n",
    "            merged = [v[0]]\n",
    "            for current in v:\n",
    "                previous = merged[-1]\n",
    "                if current[0] <= previous[1]:\n",
    "                    previous[1] = max(previous[1], current[1])\n",
    "                else:\n",
    "                    merged.append(current)\n",
    "            for i in merged:\n",
    "                genes_str.write(k+\"\\t\"+str(i[0])+\"\\t\"+str(i[1])+\"\\n\")\n",
    "\n",
    "    counter+=1\n",
    "    bed_pat=patient\n",
    "                "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
